File: 150. Evaluate Reverse Polish Notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = ['+', '-', '*', '/']
        for token in tokens:
            if token not in ops:
                stack.append(int(token))
            else:
                a = stack.pop()
                b = stack.pop()
                if token == '+':
                    stack.append(b+a)
                elif token == '-':
                    stack.append(b-a)
                elif token == '*':
                    stack.append(b*a)
                elif token == '/':
                    if a*b < 0:
                        stack.append(-(b//-a))
                    else:
                        stack.append(b//a)
        return stack[-1]

File: 150. Evaluate Reverse Polish Notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        HashSet<String> ops = new HashSet<>();
        ops.add("+");
        ops.add("-");
        ops.add("*");
        ops.add("/");
        Deque<Integer> stack = new ArrayDeque<>();
        for (String t : tokens) {
            if (ops.contains(t)) {
                int n2 = stack.pop();
                int n1 = stack.pop();
                int res = 0;
                if (t.equals("+")) {
                    res = n1 + n2;
                } else if (t.equals("-")) {
                    res = n1 - n2;
                } else if (t.equals("*")) {
                    res = n1 * n2;
                } else {
                    res = n1 / n2;
                }
                stack.push(res);
            } else {
                stack.push(Integer.parseInt(t));
            }
        }
        return stack.pop();
    }
}